src/storage/mvcc/reader/scanner/entry.rs

```diff
@@ -0,0 +1,700 @@
+use crate::storage::mvcc::write::{Write, WriteType};
+use crate::storage::mvcc::Result;
+use crate::storage::txn::TxnEntry;
+use crate::storage::{Cursor, Key, Lock, Snapshot, Statistics};
+
+use super::util::CheckLockResult;
+use super::ScannerConfig;
+
+/// This struct can be used to scan keys starting from the given user key (greater than or equal).
+///
+/// Internally, for each key, rollbacks are ignored and smaller version will be tried. If the
+/// isolation level is SI, locks will be checked first.
+///
+/// Use `ScannerBuilder` to build `EntryScanner`.
+///
+/// Note: The implementation is almost the same as `ForwardScanner`, made a few
+///       adjustments to output content in each cf.
+pub struct EntryScanner<S: Snapshot> {
+    cfg: ScannerConfig<S>,
+    lock_cursor: Cursor<S::Iter>,
+    write_cursor: Cursor<S::Iter>,
+    default_cursor: Cursor<S::Iter>,
+    /// Is iteration started
+    is_started: bool,
+    statistics: Statistics,
+}
+
+impl<S: Snapshot> EntryScanner<S> {
+    pub fn new(
+        mut cfg: ScannerConfig<S>,
+        lock_cursor: Cursor<S::Iter>,
+        write_cursor: Cursor<S::Iter>,
+    ) -> Result<EntryScanner<S>> {
+        // Note: Create a default cf cursor will take key range, so we need to
+        //       ensure the default cursor is created after lock and write.
+        let default_cursor = cfg.create_cf_cursor(CF_DEFAULT)?;
+        Ok(EntryScanner {
+            cfg,
+            lock_cursor,
+            write_cursor,
+            default_cursor,
+            statistics: Statistics::default(),
+            is_started: false,
+        })
+    }
+
+    /// Take out and reset the statistics collected so far.
+    pub fn take_statistics(&mut self) -> Statistics {
+        std::mem::replace(&mut self.statistics, Statistics::default())
+    }
+
+    /// Get the next key-value pair, in forward order.
+    pub fn read_next(&mut self) -> Result<Option<TxnEntry>> {
+        if !self.is_started {
+            if self.cfg.lower_bound.is_some() {
+                // TODO: `seek_to_first` is better, however it has performance issues currently.
+                self.write_cursor.seek(
+                    self.cfg.lower_bound.as_ref().unwrap(),
+                    &mut self.statistics.write,
+                )?;
+                self.lock_cursor.seek(
+                    self.cfg.lower_bound.as_ref().unwrap(),
+                    &mut self.statistics.lock,
+                )?;
+            } else {
+                self.write_cursor.seek_to_first(&mut self.statistics.write);
+                self.lock_cursor.seek_to_first(&mut self.statistics.lock);
+            }
+            self.is_started = true;
+        }
+
+        // The general idea is to simultaneously step write cursor and lock cursor.
+
+        // TODO: We don't need to seek lock CF if isolation level is RC.
+
+        loop {
+            // `current_user_key` is `min(user_key(write_cursor), lock_cursor)`, indicating
```

Comment:
If the logic is same, I prefer to remove these comments and add a hint about where it similars to, so that we won't forget to update multiple comments. This is what we did in backward scanner.